feat(MEP): Implement the run_query method

wmak opened the pull request:
- This checks the orderby for each entity, and orders the queries so the entity with the most orderbys are first.
  - If more than one orderby contains functions throw an error since the query is now impossible
- This then uses the groupby to group the results of the up to 3 queries back together
- This also implements the count_unique function so that we can have a test across 2 entities
- Part 1 [here](https://github.com/getsentry/sentry/pull/31649)

wmak committed:
feat(MEP): Add initial framework for metric queries

- This adds a MetricsQueryBuilder, which works very similarily to our
  QueryBuilder, but with specific handlers for how metrics construct
  queries
- This MetricsQueryBuilder does not yet construct snql queries, and will
  not because table queries will require multiple queries to construct
  similar table data
  - that is, if we want [transaction, p95, count_unique(user)], we need
    a query against distributions with [transaction, p95] followed by a
    second query for [transaction, count_unique(user)] against the sets
    table
    - This is so we can maintain a sortby

getsantry[bot] committed:
style(lint): Auto commit lint changes

wmak committed:
fix: typing

wmak committed:
ref: Implement count_unique(user)

wmak committed:
feat(MEP): Implement the run_query method

- This checks the orderby for each entity, and orders the queries so the
  entity with the most orderbys are first.
  - If more than one orderby contains functions throw an error since the
    query is now impossible
- This then uses the groupby to group the results of the up to 3 queries
  back together

wmak commented on src/sentry/search/events/builder.py:
```diff
@@ -1521,6 +1522,101 @@ def get_snql_query(self) -> None:
+                if transaction_map:
+                    # This is second or third query, filter to the transactions
+                    where = self.where + [
+                        Condition(
+                            self.resolve_column("transaction"), Op.IN, list(transaction_map.keys())
+                        )
+                    ]
```
need to use groupby not transaction

wmak commented:
Moving to draft, need to rework the conditions for 2nd+ queries

wmak committed:
fix: Need to group all the values by the group by not just txn

- This also adds better tests since the previous ones were not
  sufficient

wmak committed:
Merge branch 'master' into wmak/feat/run-metrics-query

wmak committed:
ref: Woops leftover from merge conflict

k-fish commented on src/sentry/search/events/builder.py:
```diff
@@ -1552,6 +1553,127 @@ def get_snql_query(self) -> None:
+            "data": None,
+            "meta": [],
+        }
+        for query_details in [query_framework.pop(primary), *query_framework.values()]:
```
This is kind of fine, but it'd probably be clearer to just do a single request with the primary then loop through the rest imo. If that's too much repeated code, then take the block below and turn it into a function.

k-fish reviewed (approved):
This is fine, but some of the code should be split into functions for clarity either now or in the future, and it could use more tests. Should also get SnS review for how the query is being built if they have a chance.

wmak commented on tests/sentry/search/events/test_builder.py:
```diff
@@ -773,3 +804,73 @@ def get_granularity(start, end):
+                "p95(transaction.duration)",
+                "count_unique(user)",
+            ],
+            orderby="-p95(transaction.duration)",
```
Agreed, going to do set & distribution for now, will put a TODO to update when we have a function that depends on `counter`

wmak committed:
ref: Addressing PR comments

evanh commented on src/sentry/search/events/builder.py:
```diff
@@ -1552,6 +1559,138 @@ def get_snql_query(self) -> None:
+                        groupby_values.append(groupby_key)
+                    value_map[value_map_key].update(row)
+                result["meta"] += current_result["meta"]
+        result["data"] = list(value_map.values())
```
Something I thought of: metrics makes no guarantees that there will be data for every aggregation for every group by. This is a contrived example, but there could be a p95 for transaction `x` but not unique users. Is your code supposed to handle that case?

wmak replied:
yes :+1: I'll add an explicit test for this, but what should happen is that we'll get p95 for `x`, and nothing for count_unique.

wmak replied:
Ah, very good point, i just realized for this scenario if we sort by unique users we'll get an unexpected result (`x` just won't show up at all), and I don't think there's a workaround either 🤔 Going to document this as a known deficiency for now.

evanh reviewed (commented):
This idea makes sense. Couple questions. Also I don't know if you planned to add this now or were going to do it later, but I would think about how to instrument this, particularly the results aggregation.

wmak commented on src/sentry/search/events/builder.py:
```diff
@@ -1552,6 +1559,138 @@ def get_snql_query(self) -> None:
+                                ],
+                            ),
+                            Op.IN,
+                            Function("tuple", groupby_values),
```
Good point, We probably don't want to allow doing the 10k limit on these table queries. I'll enforce a 51 limit on this builder for now since we only plan to use this for the performance table :+1:

evanh replied:
Also keep in mind there is a maximum size of the actual query you can build. So if you have 100k conditions, the SQL string will be too large for Clickhouse to process.

wmak commented:
re: instrumenting, I'll tackle that when its hooked up to an endpoint. I find it easier to do at that point since I'll be able to see the spans in context

wmak committed:
feat: Addressing comments

- Adding a max limit to the metric query builder since we don't ever
  want to do the group by filtering on more than a reasonable number
  - Picking 51 for now since that aligns with tables in the UI (+1 for
    pagination)
- Adding a test & skipped test with note for the case where the data is
  sparse between the tables (ie. only in distribution but not set)

Zylphrex commented on tests/sentry/search/events/test_builder.py:
```diff
@@ -773,3 +846,193 @@ def get_granularity(start, end):
+        result = query.run_query("test_query")
+        assert len(result["data"]) == 2
+        assert result["data"][0] == {
+            "transaction": indexer.resolve("foo_transaction"),
```
The result here is an integer, right? Should it be mapped back to the string for the response?

wmak replied:
TBH, still haven't decided if that makes more sense in a post-processing step, or here in the `run_query` function. Going to leave this for now and revisit when I connect this to an endpoint.

evanh commented on src/sentry/search/events/builder.py:
```diff
@@ -1552,6 +1573,139 @@ def get_snql_query(self) -> None:
+                )
+                current_result = raw_snql_query(
+                    query,
+                    referrer,
```
I thought about this and I think I would do a separate referrer to delineate between the "primary" and secondary queries, since the conditions will change between those two types of queries.

wmak committed:
ref: Updating referrer, removing accidental assert False

wmak committed:
Merge branch 'master' into wmak/feat/run-metrics-query
